### PR TITLE
Update ReefMod interface to match new species list

### LINKED ISSUE #399: Remove coral cover conversion for ReefModEngine interface
These lines need to be adjusted or removed as we're now representing the same species as ReefMod Engine:

https://github.com/open-AIMS/ADRIA.jl/blob/f906595d83de3e54df0f278d164413abe49b807e/src/ExtInterface/ReefMod/scenarios.jl#L9-L26

### PR DESCRIPTION
**Pull request by Rosejoycrocker**
Previously, the ReefMod extension required a change to initial covers as the species did not match those in ADRIA. Now that ADRIA has 6 species, the change is no longer required.

Resolves #399 .

### CODE REVIEW

File: src/ExtInterface/ReefMod/scenarios.jl
Comment (ConnectedSystems) [inline src/ExtInterface/ReefMod/scenarios.jl:6]: Hi Rose,

On reviewing what's left of the code, I think we don't need this entire function anymore?

It looks like we can call `run_scenarios()` as normal and it'll just work

Comment (Rosejoycrocker) [reply src/ExtInterface/ReefMod/scenarios.jl:6]: Sure, will do

Comment (Rosejoycrocker) [reply src/ExtInterface/ReefMod/scenarios.jl]: Removed the scenarios.jl file

Comment (Rosejoycrocker) [reply src/ExtInterface/ReefMod/scenarios.jl]: So I tried loading the Domain and running scenarios.  Domain runs fine but there's a dimension mismatch when the scenarios are run so I'll look into fixing that before this is merged.

**Comment by Rosejoycrocker** [comment]
Hi @ConnectedSystems. The dimension mismatch issue seems to be becuase the initial cover cover file for ReefMod has dimensions sites by number of species (nsites*6), whereas ADRIA's initial coral structure is sites by species and size classes (nsites*36). Do you know if the ReefMod initial data is aggregated across size classes (so we could split evenly across size classes for simplicity)? Otherwise I can ask Arne for more details.

**Comment by ConnectedSystems** [comment]
This is weird, because this was working fine before. Could you please point me to where (file and line number) this error is raised?

**Comment by Rosejoycrocker** [comment]
> This is weird, because this was working fine before. Could you please point me to where (file and line number) this error is raised?

It occurs for me at line 312 in `src/scenario.jl`, using the rme_ml_2023_03_30b domain.

**Comment by ConnectedSystems** [comment]
Line numbers don't seem to line up with anything interesting.

Which branch are you on? Have you pulled the most recent changes?

**Comment by Rosejoycrocker** [comment]
Sorry it's line 311 without an added breakpoint : https://github.com/open-AIMS/ADRIA.jl/blob/22f6a83449c481c28a9aeaa239222a63aaed1979/src/scenario.jl#L311

**Comment by ConnectedSystems** [comment]
Hi Rose,

Please update `load_initial_cover(ReefModDomain, ...)` to do the necessary transpose. This was handled in `run_scenarios()` previously (which was poor design now that I think about it).

**Comment by Rosejoycrocker** [comment]
Thanks Takuya, I've updated to transpose the data but there's still a dimension mismatch becasue the data is 6*nsites instead of 36*nsites. I think the previous code dealt with this by repeating the data but I'm not sure if this is appropriate for the ReefMod data- do you know if it is an aggregate of all size classes and should be split across size classes, or is it an average and so should be repeated across size classes? I can ask Arne if we're unsure.

**Comment by ConnectedSystems** [comment]
Yes you're right - the values were interpolated previously.

I'm not sure how ReefMod handles this - they have 10 size classes I believe. 

For now, I would split it evenly across the six size classes we have, but yes, could you check with Arne or Ken?

**Comment by Rosejoycrocker** [comment]
Thanks Takuya, I'll split evenly across size classes for now and I'll email Ken or Arne to check.

**Comment by Rosejoycrocker** [comment]
Hi @ConnectedSystems , I asked Yves-Marie and Arne about how the files coral_spX_2022.csv are used to initialise size classes in ReefMod . A lognormal distribution is used to generate the area for different size classes, so that the total area adds up to the coral cover in coral_spx_2022.csv. Maybe our equivalent to this could be to weight each size class with weights which sum to one when we do the splitting of cover across size classes. Weights could be generated from a log-normal distribution where we choose the shape parameters to best match those of ReefMod. What do you think?

**Comment by ConnectedSystems** [comment]
Is it not possible/sensible to use the same shape parameters?

**Comment by Rosejoycrocker** [comment]
We woulduse the same shape parameters but they have more size classes so we'd have to split it up differently.

**Comment by ConnectedSystems** [comment]
Makes sense to me, thank you

**Comment by Rosejoycrocker** [comment]
I just asked Yves-Marie for where the shape parameters are for ReefMod and then I can implement this :)

**Comment by Rosejoycrocker** [comment]
Hi @ConnectedSystems , Yves-Marie got back to me saying that the default distribution they use to initialise the size distribution is a log-normal distribution with mean log(700cm^2) and std log(4). What I've implemented takes the colony area at each of the bin edges in ADRIA's size classes and finds the density between each bin edge. These densities are then normalised and used to split the coral cover for each of the species into the size classes for ADRIA. An alternative would be to treat the largest size class as 60cm diameter and above and just integrate the density to infinity for the largest size class so you don't have to normalise. Let me know what you think is best.

The matrix reshaping I've used is a little convoluted so I'll try and think of another way to do it, but besides from that it's ready to review.

File: src/ExtInterface/ReefMod/Domain.jl
Comment (ConnectedSystems) [inline src/ExtInterface/ReefMod/Domain.jl:316]: Please use sentence case for comments that are not inline.

Best practice is to comment not what is being done but _why_.
The occasions you include the what is when the code is necessarily complicated.

The comment for the code here for example is unnecessary. You should write more readable code (use informative variable names) instead.

```julia
# get bin edges for coral diameters and transform to coral areas in cm^2
bin_edges_area = colony_mean_area(Float64[0, 2, 5, 10, 20, 40, 80])
```

File: src/ExtInterface/ReefMod/Domain.jl
Comment (ConnectedSystems) [inline src/ExtInterface/ReefMod/Domain.jl:319]: Why not call `cdf()` for all of `bin_edges_area` once?

File: src/ExtInterface/ReefMod/Domain.jl
Comment (ConnectedSystems) [inline src/ExtInterface/ReefMod/Domain.jl:321]: As with my previous comment - we can see from your variable name what this is - no need write a comment that repeats the same thing as the code.

File: src/ExtInterface/ReefMod/Domain.jl
Comment (ConnectedSystems) [inline src/ExtInterface/ReefMod/Domain.jl:329]: Agree this isn't the best approach, but I'll leave it to you to tidy up.

Is it not possible to apply the weights with L 325?

File: src/ExtInterface/ReefMod/Domain.jl
Comment (ConnectedSystems) [inline src/ExtInterface/ReefMod/Domain.jl:327]: As discussed previously, use the more general "locations" instead of "sites" please.

ReefMod does not work with sites.

```suggestion
    # Repeat species over each size class and reshape to give ADRIA compatible size (36 * n_locs).
```

**Review by ConnectedSystems — CHANGES_REQUESTED** [review]

File: src/ExtInterface/ReefMod/Domain.jl
Comment (Rosejoycrocker) [reply src/ExtInterface/ReefMod/Domain.jl:329]: I tried doing this previously but you need the repeats of the coral species so that you get each of the size classes for each coral species, otherwise you just get one size class for each species. I could move the repeat up and then the final line is just reshaping and multiplying by weights...

Comment (ConnectedSystems) [reply src/ExtInterface/ReefMod/Domain.jl]: So `icc_data` and `size_class_weights` are not of compatible shapes?

Comment (Rosejoycrocker) [reply src/ExtInterface/ReefMod/Domain.jl]: icc data is n_locsX6 and size_class_weights is 6X1, but if you just multiply them together you get (species x* size class x) but not (species x * size class y). You need all permutations of  (species x * size class y), for each location, to get the final 36 species/size class dimension.

Comment (ConnectedSystems) [reply src/ExtInterface/ReefMod/Domain.jl]: How is `size_class_weights` of shape 6x1? Is it not a vector?

In any case:

```julia
n_locs = 50
n_species = 6
n_sc = 6
icc_data = rand(n_locs , n_species)  # 50 example locations
size_class_weights = rand(n_sc)

# Multiply each row of icc_data with weights with each element in size_class_weights
# resulting in a matrix (shape: 50 x 6) where each entry is a vector
# You could replace icc_data instead of assigning to x
x = icc_data .* [size_class_weights]

# You can verify correctness with:
all(icc_data[1, 1] .* size_class_weights .== x[1,1])

# Join vectors for each row and create matrix
Matrix(transpose(hcat(reduce.(vcat, eachrow(x))...)))

# ... or with pipes:
# hcat(reduce.(vcat, eachrow(x))...) |> transpose |> Matrix
```

Is that what you want?

Comment (Rosejoycrocker) [reply src/ExtInterface/ReefMod/Domain.jl]: Thanks Takuya, that's perfect. I didn't realise elementwise multiplication with square brakets gives a deconstructed dot product. I'll implement this and then it's ready for review.

Comment (Rosejoycrocker) [reply src/ExtInterface/ReefMod/Domain.jl]: Ready for review :)

File: src/ExtInterface/ReefMod/Domain.jl
Comment (ConnectedSystems) [inline src/ExtInterface/ReefMod/Domain.jl:317]:
```suggestion
    # Use integral density between bounds of each size class areas to create weights for each size class
```

File: src/ExtInterface/ReefMod/Domain.jl
Comment (ConnectedSystems) [inline src/ExtInterface/ReefMod/Domain.jl:315]: Please detail what the distribution is used for (the "why")

File: src/ExtInterface/ReefMod/Domain.jl
Comment (ConnectedSystems) [inline src/ExtInterface/ReefMod/Domain.jl:331]: This isn't reordering dimensions is it?

**Comment by ConnectedSystems** [comment]
One last question: Have you tried running ADRIAmod run with RME data now?

**Comment by Rosejoycrocker** [comment]
> One last question: Have you tried running ADRIAmod run with RME data now?

Yes, I ran a small number of scenarios and they run fine.

> One last question: Have you tried running ADRIAmod run with RME data now?

This is what it looks like for a small number of scenarios:
ReefMod_ADRIA_runs

**Review by ConnectedSystems — APPROVED** [review]

File: src/ExtInterface/ReefMod/Domain.jl
Comment (ConnectedSystems) [inline src/ExtInterface/ReefMod/Domain.jl:313]: That time stamp looks suspicious. If it was this month, it can't have been AEDT?

```suggestion
    # as suggested by YM (pers comm. 2023-08-08 12:55pm AEST). Distribution is used to split ReefMod initial 
```

**Pull request merged**
